src/lex.cpp

```diff
@@ -756,27 +790,52 @@ void lexer::parse_binary_number() {
   this->input_ = check_garbage_in_number_literal(input);
 }
 
-char8* lexer::check_garbage_in_number_literal(char8* input) {
-  char8* garbage_begin = input;
+void lexer::parse_octal_number(bool strict, bool is_0o) {
+  if (is_0o) assert(strict);
+  char8* input = this->input_;
+
+  while (this->is_octal_digit(*input)) {
+    ++input;
+  }
+
+  if (!is_0o) {
+    switch (*input) {
+    QLJS_CASE_DECIMAL_DIGIT:
+      this->input_ = input;
+      this->parse_number();
+      return;
+    }
+  }
+
+  if (is_0o) switch (*input)
+    case u8'n':
+      ++input;
+
+  const char8* garbage_begin = input;
   for (;;) {
     switch (*input) {
+    // if we see a DECIMAL_DIGIT at this point it either means we're in
+    // strict mode or following another number identifier (such as a `.`),
+    // both of which are invalid
     QLJS_CASE_DECIMAL_DIGIT:
-    QLJS_CASE_IDENTIFIER_START:
-      input += 1;
+      ++input;
       break;
+    QLJS_CASE_IDENTIFIER_START:
+    case u8'.':
+      ++input;
+      goto done_parsing_garbage;
```

Comment:
Why do we stop at the first garbage character? This causes `0o123abcdef` to parse as `0o123a` (number) and `bcdef` (identifier) (with an error on `a`), right?

Reply:
Fixed